inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/pooling_gpu_b_fs_zyx_fsv16_imad.cl

```diff
@@ -44,6 +55,233 @@ inline ACCUMULATOR_TYPE FUNC(apply_pooling)(ACCUMULATOR_TYPE tmp, ACCUMULATOR_TY
+                        result = FUNC_CALL(apply_pooling16)(result, TO_ACCUMULATOR_VEC16(ch16_data));
+#if INPUT0_FEATURE_NUM % FEATURE_SLICE_SIZE != 0
+                    } else {
+                        __attribute__((opencl_unroll_hint(INPUT0_FEATURE_NUM % FEATURE_SLICE_SIZE)))
+                        for (uint k = 0; k < INPUT0_FEATURE_NUM % FEATURE_SLICE_SIZE; k++) {
+                            result[k] = FUNC_CALL(apply_pooling)(result[k], input[input_idx + k]);
+                        }
+                    }
+#endif
+
+#ifdef DYNAMIC_KERNEL_DIVIDER
+                    num_elements++;
+#endif
+                }
+            }
+        }
+    }
+#ifdef DYNAMIC_WITH_PADDING_KERNEL_DIVIDER
+    const int wend = min(offset_x + POOL_SIZE_X, INPUT0_SIZE_X + PADDING_SIZE_X);
+    const uint num_elements = (wend - offset_x) * INPUT0_SIZE_Z * INPUT0_SIZE_Y;
+#endif
+#else  // !CHECK_BOUNDRY
+    __attribute__((opencl_unroll_hint(Z_LOAD)))
+    for (uint pz = 0; pz < Z_LOAD; pz++) {
+        __attribute__((opencl_unroll_hint(Y_LOAD)))
+        for (uint py = 0; py < Y_LOAD; py++) {
+#if INPUT0_DIMS == 4
+            uint y_ind = y * Y_LOAD + py;
+            uint input_idx = INPUT0_GET_INDEX(b, f, y_ind, offset_x);
+            if (y_ind >= INPUT0_SIZE_Y)
+                continue;
+#else
+            uint z_ind = z * Z_LOAD + pz;
+            uint y_ind = y * Y_LOAD + py;
+            uint input_idx = INPUT0_GET_INDEX(b, f, z_ind, y_ind, offset_x);
+            if (z_ind >= INPUT0_SIZE_Z || y_ind >= INPUT0_SIZE_Y)
+                continue;
+#endif
+            __attribute__((opencl_unroll_hint(POOL_SIZE_X)))
+            for (uint px = 0; px < POOL_SIZE_X; px++) {
+                IN_VEC16 ch16_data;
+#if INPUT0_FEATURE_NUM % FEATURE_SLICE_SIZE != 0
+                if (!last_in_f_group) {
+#endif
+                    ch16_data = AS_TYPE(IN_VEC16, vload4(0, (__global int*)(input + input_idx)));
+                    result = FUNC_CALL(apply_pooling16)(result, TO_ACCUMULATOR_VEC16(ch16_data));
+#if INPUT0_FEATURE_NUM % FEATURE_SLICE_SIZE != 0
+                } else {
+                    __attribute__((opencl_unroll_hint(INPUT0_FEATURE_NUM % FEATURE_SLICE_SIZE)))
+                    for (uint k = 0; k < INPUT0_FEATURE_NUM % FEATURE_SLICE_SIZE; k++) {
+                        result[k] = FUNC_CALL(apply_pooling)(result[k], input[input_idx + k]);
+                    }
+                }
+#endif
+                input_idx += IN_X_PITCH;
+            }
+        }
+    }
+#if defined(DYNAMIC_KERNEL_DIVIDER) || defined(DYNAMIC_WITH_PADDING_KERNEL_DIVIDER)
+    const uint num_elements = POOL_SIZE_X * INPUT0_SIZE_Z * INPUT0_SIZE_Y;
+#endif
+#endif
+    lg_storage[loc_id] = result;
+
+    barrier(CLK_LOCAL_MEM_FENCE);
+    if (loc_id == 0) {
+        __attribute__((opencl_unroll_hint(LWS)))
+        for (uint i = 1; i < LWS; ++i) {
+            result = FUNC_CALL(apply_pooling16)(result, lg_storage[i]);
+        }
```

Comment:
Since `LWS` aligned to 16 it may contains leftovers, so I think we should skip them here to avoid final result corruption